docs/4.x/system-check.md

```diff
@@ -0,0 +1,137 @@
+
+## Types of system checks 
+
+Matomo typically differentiates between three kind of checks:
+
+### A required check
+
+A new check can be useful if your plugin requires for example a certain PHP extension or other capabilities for the plugin to work. If a Matomo installation does not have the required capability, then the user will be able to find out about this problem and find help to fix this issue.
+
+### An optional check.
+
+An optional check can be useful if there's a way to improve the performance of your plugin, or if there are other optional or recommended changes that improve the experience.
+
+### An informational check.
+
+Adding an informational output is typically useful for the plugin developer themselves to troubleshoot issues. Such information can then be easily requested from the user when a bug report is made. The user can then provide this information by simply copy / pasting the system check.
+
+Typical information could be version numbers of installed packages or information about stored data.
+
+## Creating a new system check
+
+To create a new system check, use the [console](/guides/piwik-on-the-command-line) to execute below command (requires Matomo 4.4.1):
+
+```bash
+./console generate:system-check
+```
+
+The command will ask for your plugin name and for the name of the system check you want to create. You should enter some meaningful name that describes the check well. For example, "JS Directory Write Permission Check".
+
+Once all information is provided, a system check class will be created in the `Diagnostic` directory of your plugin containing an example on how to define a system check. In above example it would have created a file `Diagnostic/JsDirectoryWritePermissionCheck.php` within the chosen plugin.
+
+### Making the new system check known to Matomo
+
+Matomo won't discover your newly created system check automatically. You have to tell Matomo that it exists using the `config/config.php` file within your plugin. 
+
+First you need to decide if your system check is of type required, optional or informational. Then you can register one or multiple checks like below:
+
+```php
+return [
+    'diagnostics.required' => array(
+        // adds a required system check
+        DI\get('Piwik\Plugins\MyPluginName\Diagnostic\JsDirectoryWritePermissionCheck'),
+    ),
+    'diagnostics.optional' => array(
+        // adds an optional system check
+        DI\get('Piwik\Plugins\MyPluginName\Diagnostic\JsDirectoryWritePermissionCheck'),
+    ),
+    'diagnostics.informational' => array(
+        // adds an informational system check
+        DI\get('Piwik\Plugins\MyPluginName\Diagnostic\JsDirectoryWritePermissionCheck'),
+    )
+];
```

Comment:
Even though it seems to work without the `DI\add`, according to the PHP DI docs, the array would be overwritten  without it. So guess we should add it...

```suggestion
return [
    'diagnostics.required' => DI\add(array(
        // adds a required system check
        DI\get('Piwik\Plugins\MyPluginName\Diagnostic\JsDirectoryWritePermissionCheck'),
    )),
    'diagnostics.optional' => DI\add(array(
        // adds an optional system check
        DI\get('Piwik\Plugins\MyPluginName\Diagnostic\JsDirectoryWritePermissionCheck'),
    )),
    'diagnostics.informational' => DI\add(array(
        // adds an informational system check
        DI\get('Piwik\Plugins\MyPluginName\Diagnostic\JsDirectoryWritePermissionCheck'),
    ))
];
```

Reply:
good one 👍